# Rapport TP1 Métaheuristiques d'optimization 

### Introduction:

Ce TP consiste à comparer la performance et l'efficacité de trois métaheuristiques du type Monte-Carlo. 
J'ai essayé de regrouper tout le code dans un seul notebook même si le développement de chaque algorithme est fait dans un fichier séparé accompagnant ce rapport. 
Ce notebook représente le rapport du travail et comporte tout le code source et le résultat de l'exécution avec des commentaires et des explications.
Pour les deux algorithmes du hill climbing et du recuit simulé, il se peut que l'algorithme rentre dans une boucle de très longue durée lors de la génération du nouveau candidat à cause des contraintes, il serait utile donc d'interrompre le Kernel et réexécuter l'appel de l'algorithme.

Deux bibliothèques seulement sont utilisées pour ce travail :

In [87]:
import numpy as np
import pandas as pd

La fonction 'bornes' permet de retourner les bornes des variables du problème et parce que les bornes des variables x3 et x4 dépendent des valeurs de z1 et z2, on doit calculer les bornes à chaque fois en fonction du nouveau point qu'on a.

Ces bornes n'étaient pas explicites, il fallait donc simplifier les différentes inégalités qu'on a dans le problème. Dans ce qui suit la démonstration :

•	x1 <= 99 : c'est une donnée, mais j'ai choisi comme borne inferieure 20 après plusieurs exécutions ou l'algorithme tournait dans une boucle très longue pour trouver un candidat acceptable.

•	x2 >= 1 : donnée, et j'ai choisi comme borne supérieure 90 pour la même raison

•	0.00954*x3 <= z2 & 0.0193*x3 <= z1 & x3 <= 200 ce qui implique que x3 <= min(0.625*x[1]/0.00954,0.625*x[0]/0.0193, 200), et j'ai choisi comme borne inferieure 1

•	-pi*x3**2 - 4/3*pi*x3**3 <= -1296000  & x3 >=10 implique que x3 >= 1296000/pi*x3**2 - 4/3*x3 (x3!=0) & x3 >= 10

ce qui implique que : x3 >= max(1296000/pi*x3**2 - 4/3*x3, 10) et on a x3 <= 240


In [88]:
def bornes(x):
    return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])

La fonction objective :

In [89]:
def cout_soudure(x):
    return 1.7781*0.625*x[1]*x[2]**2+ 0.6224*0.625*x[0]*x[2]*x[3] + 3.1661*(0.625*x[0])**2*x[3] + 19.84*(0.625*x[0])**2*x[2]


Cette fonction permet de vérifier si le nouveau point génère respecte les contraintes ou pas en se servant de la fonction 'bornes'. 

In [90]:
def contraintes(x):
    _bornes = bornes(x)
    for i in range(len(x)) :
        if x[i] < _bornes[i,0] or x[i] > _bornes[i,1]:
            return False
    return True


La fonction d'initialisation du premier point pour les deux algorithmes du hill climbing et du recuit simulé et qui le génère en respectant les contraintes imposées en se servant de la fonction 'bornes'. 

In [91]:
def initialiser_points():
    x = np.zeros(4)
    _bornes = bornes(x)
    x[0]= np.random.uniform(high = _bornes[0,1])
    x[1]= np.random.uniform(low =  _bornes[1,0], high =  _bornes[1,1])
    _bornes = bornes(x)
    x[2]= np.random.uniform(high =  _bornes[2,1])
    _bornes = bornes(x)
    x[3]= np.random.uniform(low =  _bornes[3,0], high =  _bornes[3,1])

    return x



Cette fonction est pour générer une liste de n points pour l'algorithme aléatoire.

In [92]:
def generer_points(n):
    x = np.zeros([n,4])
    for i in range(n):
        _bornes = bornes(x[i])
        x[i,0]= np.random.uniform(high = _bornes[0,1])
        x[i,1]= np.random.uniform(low =  _bornes[1,0], high =  _bornes[1,1])
        _bornes = bornes(x[i])
        x[i,2]= np.random.uniform(high =  _bornes[2,1])
        _bornes = bornes(x[i])
        x[i,3]= np.random.uniform(low =  _bornes[3,0], high =  _bornes[3,1])

    return x
   


La fonction 'mutation' est utilisée aussi pour les deux algorithmes du hill climbing et du recuit simulé. Elle permet de modifier une des variables choisies aléatoirement du candidat actuel. Elle retourne un nouveau point en fonction de l'ancien à condition qu'il respecte les contraintes.
Pour ce, j'ai utilisé une mutation trouvée dans une référence du cours qui se sert de la distribution gaussienne mais j'ai modifié la constante de réduction (0.01 ==> 0.1) parce que l'algorithme prenait un temps infini pour trouver un nouveau candidat acceptable, et l'augmentation de cette constante permet plus d'excitation du point et ne pas rester coince dans un petit intervalle. En plus, si la fonction ne trouve pas un nouveau candidat acceptable pendant dix itérations, elle change la variable à modifier.


In [93]:
def mutation(candidat):
    _bornes = bornes(candidat)

    
    temp_array = candidat

    cpt=0
    while True: #do-while
        
        if(cpt%10 == 0):
            i = np.random.randint(0,4)
            temp=candidat[i]
        
        temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0]) 
        temp_array[i] = temp

        if contraintes(temp_array):
            candidat[i] = temp
            break
        
        cpt+=1
    return candidat


Les différentes heuristiques de refroidissement pour le recuit simulé:

In [94]:
def calculer_temperature_expo(t, T_init,epsilon_t=0.025): #exponentiel
    return (1-epsilon_t)**t*T_init



In [95]:
def calculer_temperature_log(t, T_init): # logarithmique
    if t<3:
        return T_init
    return T_init/np.log(t)


In [96]:
def calculer_temperature_poly(t,T_init,t_max=200, alpha=2): # polynomial
    return (1-t/t_max)**alpha * T_init # t_max est l'iteration maximale pour laquelle la temperature doit etre nulle


Fonction de calcul de la probalbilite d'acceptation d'un nouveau point dont le score est plus mauvais que celui du point atuel.

In [97]:
def proba_acceptance(T,deltaE):
    return np.exp(-deltaE/T)

'print_results' permet d'afficher les meilleurs résultats trouves. 

In [98]:
def print_results(hist):
    
    scores = hist[:,1]
    best_score = min(scores)
    index = np.where(scores == best_score)
    #print(f"index : {index}")
    print(f"meilleurs dimensions : \n z1 = {0.625*hist[index[0][0],3]} \n z2 = {0.625*hist[index[0][0],4]} \n x3 = {hist[index[0][0],5]} \n x4 = {hist[index[0][0],6]}\n")
    print(f"meilleur score : {best_score}\n")

   



Cette fonction prepare les donnees pour la presentation dans un tableau a l'aide de la bibliotheque 'pandas'. Elle permet de convertir une liste numpy en une DataFrame, avec le calcul de z1 et z2 a partir de x1 et x2.

In [99]:
def preparation_table(hist):
    hist[:,3] = [0.625 * a for a in hist[:,3]]
    hist[:,4] = [0.625 * a for a in hist[:,4]]

    table = pd.DataFrame(data=hist[:,1:], columns=["Score","Nbr itrs", "z1","z2","x3","x4"], index=hist[:,0]) 
    
    return table




Fonction pour l'affichage de l'historique détaillé des points et scores.

In [100]:
def print_table(hist):
      print("historique detaille :\n")
      return preparation_table(hist)

Fonction pour l'affichage d'une table contenant l'analyse de l'historique à l'aide de la méthode describe() de 'pandas'.

In [101]:
def print_analyse(hist):
    print("Analyse des donnees generees :\n")
    return preparation_table(hist).describe()


L'algorithme de la recherche aléatoire : 

In [154]:
def aleatoire(n = 50, itrs=100,epsilon = 0.001, samples_size = 10 ,verbose = False):
    
     
       #iterateur 
    i=0
    historique = np.empty([0,7])
    while(i<n):
        i+=1
        k=0
        l=0
        current_points = generer_points(samples_size)
        scores = np.apply_along_axis(cout_soudure, arr = current_points, axis = 1)
        score = np.min(scores)
        current_point = current_points[np.argmin(scores),]
        best_point = current_point
        best_score = score
        best_before_score = best_score
        
        while(k<itrs):
          k+=1   
          current_points = generer_points(samples_size)
          scores = np.apply_along_axis(cout_soudure, arr = current_points, axis = 1)

          score = np.min(scores)
          current_point = current_points[np.argmin(scores),]          
          
          best_before_score = best_score

          if(score < best_score):
                best_score = score
                best_point = current_point
          

          if best_before_score - best_score < epsilon :
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              break
          

          if verbose :
                print(f"itr : {k}, minimum : {current_point}, best score : {score}")
        
        historique = np.append(historique,[np.append([i,best_score,k],[best_point])],axis=0)
    return historique

In [155]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Veuillez reesayer!")
    
hist = aleatoire(itrs = int(iter), epsilon=float(epsilon))

Entrer le nombre d'iterations :  10
Entrer epsilon ( progres minimal ) :  0.01


<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [156]:
print_results(hist)

meilleurs dimensions : 
 z1 = 0.10967516169994439 
 z2 = 1.3900929424471427 
 x3 = 3.9140281888200943 
 x4 = 6508.165654562978

meilleur score : 2025.4993228878886



In [157]:
cout_soudure([0.10967516169994439 /0.625 ,1.3900929424471427 /0.625,3.9140281888200943   , 6508.165654562978])

2025.4993228878886

In [147]:
print_table(hist)

historique detaille :



Score  Nbr itrs        z1         z2         x3            x4
1.0     8739.101470      10.0  0.348447  21.177219  11.655083  1.234819e+03
2.0     1142.589584      10.0  0.449237  50.646601   2.699785  3.410628e+02
3.0    15617.474135      10.0  1.206392   3.185345  33.276868  2.833247e+02
4.0   105101.333874      10.0  7.300744   1.309975  65.237642  5.632049e+01
5.0     5333.044134      10.0  0.334448  52.265618   4.108778  3.104776e+03
6.0    43359.226424      10.0  2.702651   5.679444  50.253161  9.821608e+01
7.0    27172.174104      10.0  1.252451  11.315218  24.010052  6.260362e+02
8.0     4446.410424      10.0  0.292373  25.130292   8.862704  4.892963e+02
9.0    26890.823583      10.0  1.308901  46.087446  10.477231  1.256415e+03
10.0  118135.358169      10.0  2.283404  34.624186  40.046545  2.078266e+02
11.0    3260.269784      10.0  0.120293  19.266604   6.062480  4.001134e+03
12.0    6919.247450      10.0  0.549701  11.832892   8.414010  1.402520e+03
13.0   35579.444345      10.0  2.947995   4.431786  26.342311  3.371060e+02
14.0   37224.005175      10.0  1.386644   6.200911  51.678199  1.145544e+02
15.0    2491.906808      10.0  0.037553  42.596165   1.495794  5.890598e+04
16.0   28335.542363      10.0  2.003404  12.401649  18.503823  5.397348e+02
17.0    2426.514044      10.0  0.065571  35.383868   2.427631  1.824088e+04
18.0   41863.500305      10.0  1.039071  33.365379  21.180838  8.643226e+02
19.0  101503.200388      10.0  2.822052  17.374755  50.864629  1.181688e+02
20.0   32798.650004      10.0  0.710677  13.169366  34.336843  2.886819e+02
21.0   23627.264677      10.0  1.181348  53.230314   5.742979  2.354600e+03
22.0    5928.269331      10.0  0.082576  13.426040   4.005194  2.437935e+04
23.0    5587.844633      10.0  0.365808  53.686444   5.792878  1.359515e+03
24.0   94078.164223      10.0  3.314273  13.954727  49.695697  1.600149e+02
25.0    4831.797960      10.0  0.198186  12.725393   7.903179  3.104328e+03
26.0    2585.618951      10.0  0.112422  37.761167   4.646028  3.109055e+03
27.0    4792.809375      10.0  0.158634  23.926114   2.812331  1.246673e+04
28.0    6628.389945      10.0  1.187470  11.330306  11.052417  3.054025e+02
29.0    6259.535819      10.0  0.058676  30.175930   2.094629  6.892767e+04
30.0    5367.546689      10.0  0.322068   7.108428   7.461640  2.548311e+03
31.0   36383.278548      10.0  0.938056  47.302174  15.105497  1.458657e+03
32.0   32753.047200      10.0  2.745412  29.531136  16.231150  3.196254e+02
33.0   28697.765605      10.0  1.463763   7.171896  36.668467  2.486294e+02
34.0    9855.782666      10.0  0.328216  15.048560  12.888968  1.810014e+03
35.0   12081.771370      10.0  0.113909  36.230828   2.694442  5.003409e+04
36.0   10189.742192      10.0  0.172067  21.925096   4.302852  1.706869e+04
37.0   19956.631379      10.0  0.781469  20.862740  17.428655  8.142915e+02
38.0    6548.700794      10.0  0.350175  41.116795   3.558096  4.824602e+03
39.0   12392.806552      10.0  0.997976  38.970089  11.232645  3.383921e+02
40.0    6829.117709      10.0  0.369785  22.899663  10.126208  9.503922e+02
41.0  150478.909710      10.0  5.198477  40.202705  34.015185  2.532048e+02
42.0    9129.051385      10.0  0.386147   8.292165  19.269383  7.049123e+02
43.0   57955.950882      10.0  0.797268  22.047071   4.365146  1.367790e+04
44.0   21614.576139      10.0  0.676204   9.928832  31.439324  2.642554e+02
45.0   32748.657352      10.0  2.364576   6.454280  30.263157  3.033507e+02
46.0   15892.684195      10.0  0.609990  53.218933   9.772123  1.387862e+03
47.0    5599.836978      10.0  0.009855   7.335645   0.494226  1.676078e+06
48.0   26040.703640      10.0  1.024388  44.618135   9.570196  1.971020e+03
49.0   73204.527113      10.0  0.838558  48.829734  27.626897  3.935617e+02
50.0   28440.680632      10.0  0.623810  12.044992   2.163420  1.366963e+04

In [148]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs         z1         z2         x3            x4
count      50.000000      50.0  50.000000  50.000000  50.000000  5.000000e+01
mean    28096.425604      10.0   0.711670  15.485013  17.447148  3.995595e+04
std     33351.418918       0.0   0.864776  10.166518  16.313975  2.365565e+05
min      1142.589584      10.0   0.006160   0.818735   0.494226  5.632049e+01
25%      6011.085953      10.0   0.202253   7.159970   4.435367  3.089582e+02
50%     15755.079165      10.0   0.433400  13.469474  10.764824  1.092605e+03
75%     32787.249303      10.0   0.809243  24.167412  27.305750  3.107985e+03
max    150478.909710      10.0   4.562965  33.554027  65.237642  1.676078e+06

L'algorithme du Hill Climbing : 

In [107]:
def hill_climbing(itrs = 100,n = 50, epsilon = 0.001, verbose = False):
    i=0
    historique = np.empty([0,7])

    while(i<n):
      i+=1
      currentPoint = initialiser_points()
      bestScore = cout_soudure(currentPoint)
      bestPoint = currentPoint
      beforeBestScore = bestScore
      k=0 #iterateur 
      l=0

      while(k<itrs):

          k+=1   
          currentPoint = mutation(currentPoint)
          score = cout_soudure(currentPoint)
          
          beforeBestScore = bestScore

          if(score < bestScore):
                bestScore = score
                bestPoint = currentPoint

          if beforeBestScore - bestScore < epsilon :
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              break
          

          if verbose :
                print(f"itr : {k}, minimum : {currentPoint},  score : {score}")
    
      historique = np.append(historique,[np.append([i,bestScore,k],[bestPoint])],axis=0)

    return historique

In [109]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Veuillez reesayer!")
    
hist = hill_climbing(itrs = int(iter), epsilon=float(epsilon))

Entrer le nombre d'iterations :  10
Entrer epsilon ( progres minimal ) :  0.01


<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: overflow encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: overflow encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: invalid value encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])


In [110]:
print_results(hist)

meilleurs dimensions : 
 z1 = nan 
 z2 = nan 
 x3 = nan 
 x4 = nan

meilleur score : 11472.343330839867



In [111]:
print_table(hist)

historique detaille :



Score  Nbr itrs  z1  z2  x3  x4
1.0   2.071193e+06      10.0 NaN NaN NaN NaN
2.0   8.976210e+05      10.0 NaN NaN NaN NaN
3.0   2.905953e+05      10.0 NaN NaN NaN NaN
4.0   1.253217e+06      10.0 NaN NaN NaN NaN
5.0   1.736549e+06      10.0 NaN NaN NaN NaN
6.0   5.777901e+05      10.0 NaN NaN NaN NaN
7.0   4.031697e+06      10.0 NaN NaN NaN NaN
8.0   3.515628e+06      10.0 NaN NaN NaN NaN
9.0   1.034275e+07      10.0 NaN NaN NaN NaN
10.0  6.935583e+06      10.0 NaN NaN NaN NaN
11.0  7.118020e+06      10.0 NaN NaN NaN NaN
12.0  5.009684e+06      10.0 NaN NaN NaN NaN
13.0  5.995512e+06      10.0 NaN NaN NaN NaN
14.0  4.488927e+06      10.0 NaN NaN NaN NaN
15.0  5.961843e+06      10.0 NaN NaN NaN NaN
16.0  5.455070e+06      10.0 NaN NaN NaN NaN
17.0  6.109300e+06      10.0 NaN NaN NaN NaN
18.0  1.059845e+06      10.0 NaN NaN NaN NaN
19.0  1.147234e+04      10.0 NaN NaN NaN NaN
20.0  1.104119e+07      10.0 NaN NaN NaN NaN
21.0  1.121079e+07      10.0 NaN NaN NaN NaN
22.0  1.893723e+06      10.0 NaN NaN NaN NaN
23.0  3.703932e+06      10.0 NaN NaN NaN NaN
24.0  5.377022e+08      10.0 NaN NaN NaN NaN
25.0  6.312840e+06      10.0 NaN NaN NaN NaN
26.0  1.327072e+06      10.0 NaN NaN NaN NaN
27.0  7.442915e+06      10.0 NaN NaN NaN NaN
28.0  8.708737e+04      10.0 NaN NaN NaN NaN
29.0  4.663730e+06      10.0 NaN NaN NaN NaN
30.0  3.001768e+05      10.0 NaN NaN NaN NaN
31.0  6.735932e+05      10.0 NaN NaN NaN NaN
32.0  2.151164e+06      10.0 NaN NaN NaN NaN
33.0  1.771245e+06      10.0 NaN NaN NaN NaN
34.0  2.732555e+06      10.0 NaN NaN NaN NaN
35.0  1.165415e+06      10.0 NaN NaN NaN NaN
36.0  1.095662e+07      10.0 NaN NaN NaN NaN
37.0  5.907172e+06      10.0 NaN NaN NaN NaN
38.0  2.461085e+06      10.0 NaN NaN NaN NaN
39.0  7.556717e+06      10.0 NaN NaN NaN NaN
40.0  3.488839e+06      10.0 NaN NaN NaN NaN
41.0  1.381092e+07      10.0 NaN NaN NaN NaN
42.0  3.245469e+06      10.0 NaN NaN NaN NaN
43.0  6.627442e+05      10.0 NaN NaN NaN NaN
44.0  1.159276e+07      10.0 NaN NaN NaN NaN
45.0  7.108505e+05      10.0 NaN NaN NaN NaN
46.0  6.370629e+06      10.0 NaN NaN NaN NaN
47.0  5.254920e+06      10.0 NaN NaN NaN NaN
48.0  3.361753e+06      10.0 NaN NaN NaN NaN
49.0  1.297664e+07      10.0 NaN NaN NaN NaN
50.0  3.531646e+06      10.0 NaN NaN NaN NaN

In [112]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs   z1   z2   x3   x4
count  5.000000e+01      50.0  0.0  0.0  0.0  0.0
mean   1.517861e+07      10.0  NaN  NaN  NaN  NaN
std    7.549316e+07       0.0  NaN  NaN  NaN  NaN
min    1.147234e+04      10.0  NaN  NaN  NaN  NaN
25%    1.429441e+06      10.0  NaN  NaN  NaN  NaN
50%    3.617789e+06      10.0  NaN  NaN  NaN  NaN
75%    6.356182e+06      10.0  NaN  NaN  NaN  NaN
max    5.377022e+08      10.0  NaN  NaN  NaN  NaN

L'algorithme du recuit simulé : 
Après chaque m itérations, on ajoute un nombre aléatoire par la loi uniforme entre 0 et 1. Par défaut, m est fixe à 10 itérations.


In [113]:
def recuit_simule(T_init = 400,temperature= calculer_temperature_poly,itrs= 100,n = 50, m=10,epsilon = 0.001, verbose = False):
      # T_init ne doit pas etre posee comme ça
      
    
    historique = np.empty([0,7])
    i=0
    while(i<n):
        i+=1
        current_point = initialiser_points()
        best_point = current_point
        current_score = cout_soudure(current_point)
        best_score = current_score
        bestc_before_score = best_score
        k=0 # iterateur 
        l=0 # iterations pour suivre l'amelioration
        while(k<itrs):
          k+= 1
          
          new_point = mutation(current_point)
          new_score = cout_soudure(new_point)
          
          deltaE = new_score - current_score
          
          best_before_score = best_score

          if  deltaE <= 0 : 
              current_point = new_point
              current_score = new_score
              if current_score < best_score:
                  best_point = current_point
                  best_score = current_score
                
          else :
              T = temperature(k,T_init) + k//m * np.random.uniform(0,1) # Rechauffement apres chaque m iterations
              if np.random.uniform(0,1) < proba_acceptance(T,deltaE):
                  current_point = new_point
                  current_score = new_score


          if best_before_score - best_score < epsilon :  #peu ou pas d'amelioration 
                 l += 1 
          else :
                 l = 0

          if l == 50 : 
              current_point = best_point         #Retour au dernier meilleur point
              current_score = best_score
          

          if verbose :
                print(f"itr : {k}, point : {current_point}, score : {current_score}")
                
        historique = np.append(historique,[np.append([i,best_score,k],[best_point])],axis=0)

    return historique

Resultats pour le refroidissement polynomial : 

In [114]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(itrs = int(iter), epsilon=float(epsilon))

Entrer le nombre d'iterations :  10
Entrer epsilon ( progres minimal ) :  0.01


<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: overflow encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: invalid value encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])
<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: overflow encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])


In [115]:
print_results(hist)

meilleurs dimensions : 
 z1 = nan 
 z2 = nan 
 x3 = nan 
 x4 = nan

meilleur score : 74600.05620294383



In [116]:
print_table(hist)

historique detaille :



Score  Nbr itrs  z1  z2  x3  x4
1.0   1.428916e+08      10.0 NaN NaN NaN NaN
2.0   4.869368e+06      10.0 NaN NaN NaN NaN
3.0   2.848856e+06      10.0 NaN NaN NaN NaN
4.0   1.046325e+06      10.0 NaN NaN NaN NaN
5.0   9.022828e+06      10.0 NaN NaN NaN NaN
6.0   1.252210e+05      10.0 NaN NaN NaN NaN
7.0   6.521440e+06      10.0 NaN NaN NaN NaN
8.0   4.473460e+06      10.0 NaN NaN NaN NaN
9.0   2.087612e+06      10.0 NaN NaN NaN NaN
10.0  5.878205e+05      10.0 NaN NaN NaN NaN
11.0  1.424344e+06      10.0 NaN NaN NaN NaN
12.0  2.315884e+05      10.0 NaN NaN NaN NaN
13.0  4.471757e+06      10.0 NaN NaN NaN NaN
14.0  5.283003e+06      10.0 NaN NaN NaN NaN
15.0  5.187495e+06      10.0 NaN NaN NaN NaN
16.0  3.458752e+06      10.0 NaN NaN NaN NaN
17.0  1.139209e+06      10.0 NaN NaN NaN NaN
18.0  2.494985e+06      10.0 NaN NaN NaN NaN
19.0  5.735646e+06      10.0 NaN NaN NaN NaN
20.0  1.409056e+06      10.0 NaN NaN NaN NaN
21.0  1.410385e+08      10.0 NaN NaN NaN NaN
22.0  2.457232e+06      10.0 NaN NaN NaN NaN
23.0  1.688809e+05      10.0 NaN NaN NaN NaN
24.0  8.489294e+05      10.0 NaN NaN NaN NaN
25.0  2.312895e+06      10.0 NaN NaN NaN NaN
26.0  1.493551e+07      10.0 NaN NaN NaN NaN
27.0  6.298370e+06      10.0 NaN NaN NaN NaN
28.0  3.525496e+06      10.0 NaN NaN NaN NaN
29.0  1.684785e+07      10.0 NaN NaN NaN NaN
30.0  2.305343e+06      10.0 NaN NaN NaN NaN
31.0  2.809858e+07      10.0 NaN NaN NaN NaN
32.0  2.071286e+06      10.0 NaN NaN NaN NaN
33.0  9.350746e+05      10.0 NaN NaN NaN NaN
34.0  2.749445e+06      10.0 NaN NaN NaN NaN
35.0  2.766321e+06      10.0 NaN NaN NaN NaN
36.0  3.107209e+06      10.0 NaN NaN NaN NaN
37.0  2.556482e+06      10.0 NaN NaN NaN NaN
38.0  1.323359e+06      10.0 NaN NaN NaN NaN
39.0  1.024813e+06      10.0 NaN NaN NaN NaN
40.0  1.484142e+06      10.0 NaN NaN NaN NaN
41.0  1.214867e+07      10.0 NaN NaN NaN NaN
42.0  2.298940e+06      10.0 NaN NaN NaN NaN
43.0  4.080629e+06      10.0 NaN NaN NaN NaN
44.0  3.893700e+06      10.0 NaN NaN NaN NaN
45.0  4.473559e+06      10.0 NaN NaN NaN NaN
46.0  6.957887e+06      10.0 NaN NaN NaN NaN
47.0  7.460006e+04      10.0 NaN NaN NaN NaN
48.0  1.058292e+07      10.0 NaN NaN NaN NaN
49.0  4.143995e+06      10.0 NaN NaN NaN NaN
50.0  3.744033e+06      10.0 NaN NaN NaN NaN

In [117]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs   z1   z2   x3   x4
count  5.000000e+01      50.0  0.0  0.0  0.0  0.0
mean   9.891300e+06      10.0  NaN  NaN  NaN  NaN
std    2.767697e+07       0.0  NaN  NaN  NaN  NaN
min    7.460006e+04      10.0  NaN  NaN  NaN  NaN
25%    1.439294e+06      10.0  NaN  NaN  NaN  NaN
50%    2.978032e+06      10.0  NaN  NaN  NaN  NaN
75%    5.259126e+06      10.0  NaN  NaN  NaN  NaN
max    1.428916e+08      10.0  NaN  NaN  NaN  NaN

Resultats pour le refroidissement exponentiel : 

In [132]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(temperature = calculer_temperature_expo,itrs = int(iter), epsilon=float(epsilon))

Entrer le nombre d'iterations :  10
Entrer epsilon ( progres minimal ) :  0.01


<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: overflow encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: overflow encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: invalid value encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])


In [133]:
print_results(hist)

meilleurs dimensions : 
 z1 = nan 
 z2 = nan 
 x3 = nan 
 x4 = nan

meilleur score : 80722.41046934108



In [134]:
print_table(hist)

historique detaille :



Score  Nbr itrs  z1  z2  x3  x4
1.0   6.842398e+06      10.0 NaN NaN NaN NaN
2.0   3.953953e+06      10.0 NaN NaN NaN NaN
3.0   4.497081e+06      10.0 NaN NaN NaN NaN
4.0   7.833593e+06      10.0 NaN NaN NaN NaN
5.0   2.681470e+06      10.0 NaN NaN NaN NaN
6.0   3.202396e+05      10.0 NaN NaN NaN NaN
7.0   2.616199e+06      10.0 NaN NaN NaN NaN
8.0   1.831235e+06      10.0 NaN NaN NaN NaN
9.0   1.671030e+06      10.0 NaN NaN NaN NaN
10.0  1.769768e+06      10.0 NaN NaN NaN NaN
11.0  6.236221e+06      10.0 NaN NaN NaN NaN
12.0  1.463088e+07      10.0 NaN NaN NaN NaN
13.0  5.459978e+05      10.0 NaN NaN NaN NaN
14.0  2.535930e+06      10.0 NaN NaN NaN NaN
15.0  1.623755e+06      10.0 NaN NaN NaN NaN
16.0  1.941030e+06      10.0 NaN NaN NaN NaN
17.0  3.493211e+05      10.0 NaN NaN NaN NaN
18.0  4.879906e+06      10.0 NaN NaN NaN NaN
19.0  5.034677e+05      10.0 NaN NaN NaN NaN
20.0  7.923249e+05      10.0 NaN NaN NaN NaN
21.0  1.781023e+07      10.0 NaN NaN NaN NaN
22.0  8.072241e+04      10.0 NaN NaN NaN NaN
23.0  8.800999e+06      10.0 NaN NaN NaN NaN
24.0  5.967344e+05      10.0 NaN NaN NaN NaN
25.0  3.448092e+06      10.0 NaN NaN NaN NaN
26.0  1.698875e+07      10.0 NaN NaN NaN NaN
27.0  1.569163e+06      10.0 NaN NaN NaN NaN
28.0  1.546339e+06      10.0 NaN NaN NaN NaN
29.0  6.117570e+06      10.0 NaN NaN NaN NaN
30.0  2.048603e+06      10.0 NaN NaN NaN NaN
31.0  3.853498e+06      10.0 NaN NaN NaN NaN
32.0  8.877046e+05      10.0 NaN NaN NaN NaN
33.0  1.784407e+06      10.0 NaN NaN NaN NaN
34.0  9.310316e+05      10.0 NaN NaN NaN NaN
35.0  2.755759e+06      10.0 NaN NaN NaN NaN
36.0  3.297053e+06      10.0 NaN NaN NaN NaN
37.0  4.286630e+06      10.0 NaN NaN NaN NaN
38.0  3.832628e+05      10.0 NaN NaN NaN NaN
39.0  4.124932e+06      10.0 NaN NaN NaN NaN
40.0  8.968467e+05      10.0 NaN NaN NaN NaN
41.0  9.445952e+06      10.0 NaN NaN NaN NaN
42.0  9.800570e+06      10.0 NaN NaN NaN NaN
43.0  3.186478e+06      10.0 NaN NaN NaN NaN
44.0  6.427574e+06      10.0 NaN NaN NaN NaN
45.0  5.746505e+06      10.0 NaN NaN NaN NaN
46.0  5.696086e+06      10.0 NaN NaN NaN NaN
47.0  4.820406e+06      10.0 NaN NaN NaN NaN
48.0  1.732145e+06      10.0 NaN NaN NaN NaN
49.0  2.274501e+05      10.0 NaN NaN NaN NaN
50.0  9.579177e+06      10.0 NaN NaN NaN NaN

In [135]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs   z1   z2   x3   x4
count  5.000000e+01      50.0  0.0  0.0  0.0  0.0
mean   4.138529e+06      10.0  NaN  NaN  NaN  NaN
std    4.137312e+06       0.0  NaN  NaN  NaN  NaN
min    8.072241e+04      10.0  NaN  NaN  NaN  NaN
25%    1.552045e+06      10.0  NaN  NaN  NaN  NaN
50%    2.718614e+06      10.0  NaN  NaN  NaN  NaN
75%    5.733900e+06      10.0  NaN  NaN  NaN  NaN
max    1.781023e+07      10.0  NaN  NaN  NaN  NaN

Resultats pour le refroidissement logarithmique : 

In [140]:
while True:
    iter = input("Entrer le nombre d'iterations : ")
    epsilon = input("Entrer epsilon ( progres minimal ) : ")

    if int(iter) > 0 and float(epsilon) > 0:
      break
    else:
      print("Une des valeurs entrees n'est pas acceptee!")
    
hist = recuit_simule(temperature = calculer_temperature_log,itrs = int(iter), epsilon=float(epsilon))

Entrer le nombre d'iterations :  10
Entrer epsilon ( progres minimal ) :  0.01


<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: divide by zero encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-88-d5d27ed1eefe>:2: RuntimeWarning: overflow encountered in double_scalars
  return np.asarray([[20,99],[1,90],[1,min(0.625*x[1]/0.00954, 0.625*x[0]/0.0193, 200)],[max(10,1296000/(np.pi*x[2]**2)-4/3*x[2]),240]])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: overflow encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])
<ipython-input-93-01ceb51c518d>:14: RuntimeWarning: invalid value encountered in double_scalars
  temp = candidat[i] + np.random.normal(0,1) * 0.1 * temp * (_bornes[i,1]- _bornes[i,0])


In [141]:
print_results(hist)

meilleurs dimensions : 
 z1 = nan 
 z2 = nan 
 x3 = nan 
 x4 = nan

meilleur score : 86808.22291831818



In [142]:
print_table(hist)

historique detaille :



Score  Nbr itrs  z1  z2  x3  x4
1.0   9.558347e+06      10.0 NaN NaN NaN NaN
2.0   8.594413e+06      10.0 NaN NaN NaN NaN
3.0   2.610991e+06      10.0 NaN NaN NaN NaN
4.0   5.714483e+06      10.0 NaN NaN NaN NaN
5.0   9.259973e+06      10.0 NaN NaN NaN NaN
6.0   9.970838e+06      10.0 NaN NaN NaN NaN
7.0   8.680822e+04      10.0 NaN NaN NaN NaN
8.0   6.821470e+05      10.0 NaN NaN NaN NaN
9.0   4.319340e+06      10.0 NaN NaN NaN NaN
10.0  2.074685e+05      10.0 NaN NaN NaN NaN
11.0  8.385209e+06      10.0 NaN NaN NaN NaN
12.0  1.666427e+06      10.0 NaN NaN NaN NaN
13.0  1.110125e+07      10.0 NaN NaN NaN NaN
14.0  7.099688e+06      10.0 NaN NaN NaN NaN
15.0  9.093015e+06      10.0 NaN NaN NaN NaN
16.0  1.525634e+06      10.0 NaN NaN NaN NaN
17.0  1.994507e+06      10.0 NaN NaN NaN NaN
18.0  9.700564e+05      10.0 NaN NaN NaN NaN
19.0  1.733815e+07      10.0 NaN NaN NaN NaN
20.0  4.415822e+06      10.0 NaN NaN NaN NaN
21.0  1.676949e+06      10.0 NaN NaN NaN NaN
22.0  6.480845e+06      10.0 NaN NaN NaN NaN
23.0  1.069166e+07      10.0 NaN NaN NaN NaN
24.0  3.371535e+06      10.0 NaN NaN NaN NaN
25.0  6.650978e+06      10.0 NaN NaN NaN NaN
26.0  3.893268e+06      10.0 NaN NaN NaN NaN
27.0  4.293380e+06      10.0 NaN NaN NaN NaN
28.0  1.098661e+06      10.0 NaN NaN NaN NaN
29.0  1.628124e+05      10.0 NaN NaN NaN NaN
30.0  7.183729e+06      10.0 NaN NaN NaN NaN
31.0  1.670973e+06      10.0 NaN NaN NaN NaN
32.0  5.807070e+06      10.0 NaN NaN NaN NaN
33.0  2.099781e+07      10.0 NaN NaN NaN NaN
34.0  5.718852e+05      10.0 NaN NaN NaN NaN
35.0  9.865544e+06      10.0 NaN NaN NaN NaN
36.0  8.164301e+06      10.0 NaN NaN NaN NaN
37.0  1.581915e+10      10.0 NaN NaN NaN NaN
38.0  1.667321e+06      10.0 NaN NaN NaN NaN
39.0  1.870012e+06      10.0 NaN NaN NaN NaN
40.0  2.033220e+07      10.0 NaN NaN NaN NaN
41.0  7.900268e+05      10.0 NaN NaN NaN NaN
42.0  6.582300e+05      10.0 NaN NaN NaN NaN
43.0  1.833791e+06      10.0 NaN NaN NaN NaN
44.0  1.300749e+06      10.0 NaN NaN NaN NaN
45.0  1.276998e+07      10.0 NaN NaN NaN NaN
46.0  2.916546e+06      10.0 NaN NaN NaN NaN
47.0  5.461646e+06      10.0 NaN NaN NaN NaN
48.0  1.223138e+07      10.0 NaN NaN NaN NaN
49.0  1.080649e+07      10.0 NaN NaN NaN NaN
50.0  5.994177e+05      10.0 NaN NaN NaN NaN

In [143]:
print_analyse(hist)

Analyse des donnees generees :



Score  Nbr itrs   z1   z2   x3   x4
count  5.000000e+01      50.0  0.0  0.0  0.0  0.0
mean   3.219913e+08      10.0  NaN  NaN  NaN  NaN
std    2.236362e+09       0.0  NaN  NaN  NaN  NaN
min    8.680822e+04      10.0  NaN  NaN  NaN  NaN
25%    1.666651e+06      10.0  NaN  NaN  NaN  NaN
50%    4.367581e+06      10.0  NaN  NaN  NaN  NaN
75%    9.218233e+06      10.0  NaN  NaN  NaN  NaN
max    1.581915e+10      10.0  NaN  NaN  NaN  NaN

### Conclusion

En moyenne sur 50 exécutions des 3 algorithmes, l'algorithme aléatoire est celui qui a donné les meilleurs résultats ce qui est assez surprenant. Le deuxième meilleur résultat est celui du hill climbing avec le recuit simulé en dernière place avec ses trois variantes : refroidissement polynomial, exponentiel et logarithmique par ordre respectif. Mais cela ne reflète pas vraiment la performance générale de l'algorithme car d'un point je n'ai pas choisi un grand nombre d'itérations, et d'un autre point, les métaheuristiques du hill climbing et du recuit simulé ne sont pas parfaitement implémentes car plusieurs constantes doivent être précisées par expérimentation, ce que j'ai fait juste par intuition. Du coup, ces métaheuristiques ont encore besoin d'optimisation. J'ai choisi pour toutes les exécutions effectuées 10 itérations par cycle (50 cycles) et 0.01 pour epsilon. 
En ce qui concerne la durée d'exécution en termes de nombre d'itérations moyen des 3 algorithmes, il est le même. Normalement, on doit être capable de voir la différence dans le temps d'exécution en augmentant le nombre d'itérations.
En conclusion, le meilleur cout trouve est de 2025.4993228878886
